### ISSUE TITLE
Fast reloads leave orphaned processes

### ISSUE THREAD

**Issue by lloesche** [haproxy/haproxy#48]
When reloading haproxy too fast on EL7 (RedHat, CentOS) the system is being filled with orphaned processes.

I encountered this problem on CentOS 7 with haproxy-1.5.4-4.el7_1.x86_64 but expect it to exist on all systems using haproxy-systemd-wrapper not just those based on Fedora.

Steps to reproduce:

1) haproxy is running normal.
```
[root@localhost ~]# ps ax | grep haproxy
 3140 ?        Ss     0:00 /usr/sbin/haproxy-systemd-wrapper -f /etc/haproxy/haproxy.cfg -p /run/haproxy.pid
 3141 ?        S      0:00 /usr/sbin/haproxy -f /etc/haproxy/haproxy.cfg -p /run/haproxy.pid -Ds
 3142 ?        Ss     0:00 /usr/sbin/haproxy -f /etc/haproxy/haproxy.cfg -p /run/haproxy.pid -Ds
```

2) Several reloads are executed in quick succession. Problem worsens when processes happen to execute a reload in parallel.
```
[root@localhost ~]# while :; do systemctl reload haproxy; done
^C
```

3) There's multiple haproxy processes running that will never end. As you can see there's duplicate pids for the `-sf` arg. Maybe caused by a race between haproxy-systemd-wrapper reading and the new haproxy process writing it's pid.
```
[root@localhost ~]# ps ax | grep haproxy
  423 ?        S      0:00 /usr/sbin/haproxy -f /etc/haproxy/haproxy.cfg -p /run/haproxy.pid -Ds -sf 419
  429 ?        S      0:00 /usr/sbin/haproxy -f /etc/haproxy/haproxy.cfg -p /run/haproxy.pid -Ds -sf 425
  430 ?        Ss     0:00 /usr/sbin/haproxy -f /etc/haproxy/haproxy.cfg -p /run/haproxy.pid -Ds -sf 419
  431 ?        Ss     0:00 /usr/sbin/haproxy -f /etc/haproxy/haproxy.cfg -p /run/haproxy.pid -Ds -sf 425
31833 ?        Ss     0:01 /usr/sbin/haproxy-systemd-wrapper -f /etc/haproxy/haproxy.cfg -p /run/haproxy.pid
36593 ?        S      0:00 /usr/sbin/haproxy -f /etc/haproxy/haproxy.cfg -p /run/haproxy.pid -Ds -sf 36587
36600 ?        Ss     0:00 /usr/sbin/haproxy -f /etc/haproxy/haproxy.cfg -p /run/haproxy.pid -Ds -sf 36587
38316 ?        S      0:00 /usr/sbin/haproxy -f /etc/haproxy/haproxy.cfg -p /run/haproxy.pid -Ds -sf 38311
38324 ?        Ss     0:00 /usr/sbin/haproxy -f /etc/haproxy/haproxy.cfg -p /run/haproxy.pid -Ds -sf 38311
38344 ?        S      0:00 /usr/sbin/haproxy -f /etc/haproxy/haproxy.cfg -p /run/haproxy.pid -Ds -sf 38325
38350 ?        Ss     0:00 /usr/sbin/haproxy -f /etc/haproxy/haproxy.cfg -p /run/haproxy.pid -Ds -sf 38325
39667 ?        S      0:00 /usr/sbin/haproxy -f /etc/haproxy/haproxy.cfg -p /run/haproxy.pid -Ds -sf 39662
39674 ?        Ss     0:00 /usr/sbin/haproxy -f /etc/haproxy/haproxy.cfg -p /run/haproxy.pid -Ds -sf 39662
```

**Comment by dkorunic** [comment]
That's normal -- reloads keep HAProxy process handling old connections by design and if those are from clients with keepalives etc., they'll keep being up and alive.

**Comment by lloesche** [comment]
@dkorunic Hmm, I know how `-sf` makes haproxy handle old connections. But the example above is on a local VM with zero clients. The haproxy processes never end. The system will just keep piling up with processes until it eventually runs out of resources.

I'm not saying it's a good use case to reload haproxy several times a second. All I'm saying is there's a race condition in the way it gets reloaded via haproxy-systemd-wrapper.

Note this is not the case when reloading via haproxy directly. E.g. when I run
```
$ /usr/sbin/haproxy -f /etc/haproxy/haproxy.cfg -p /run/haproxy.pid -D
```
followed by
```
$ while :; do /usr/sbin/haproxy -f /etc/haproxy/haproxy.cfg -p /run/haproxy.pid -D -sf $(< /run/haproxy.pid); done
^C
```
the system is still running only one haproxy process.

You can however produce the same effect when running
```
while :; do (/usr/sbin/haproxy -f /etc/haproxy/haproxy.cfg -p /run/haproxy.pid -D -sf $(< /run/haproxy.pid)&); done
```
which doesn't wait for haproxy to return and just executes many reloads in parallel.

Now look at the following example:
```
[root@localhost ~]# ps ax | grep systemd-wrapper
33229 ?        Ss     0:00 /usr/sbin/haproxy-systemd-wrapper -f /etc/haproxy/haproxy.cfg -p /run/haproxy.pid
33463 pts/0    S+     0:00 grep --color=auto systemd-wrapper

[root@localhost ~]# while :; do kill -USR2 33229; done
^C

[root@localhost ~]# ps ax | grep haproxy
33229 ?        Ss     0:01 /usr/sbin/haproxy-systemd-wrapper -f /etc/haproxy/haproxy.cfg -p /run/haproxy.pid
33470 ?        S      0:00 /usr/sbin/haproxy -f /etc/haproxy/haproxy.cfg -p /run/haproxy.pid -Ds -sf 33458
33471 ?        S      0:00 /usr/sbin/haproxy -f /etc/haproxy/haproxy.cfg -p /run/haproxy.pid -Ds -sf 33458
33477 ?        Ss     0:00 /usr/sbin/haproxy -f /etc/haproxy/haproxy.cfg -p /run/haproxy.pid -Ds -sf 33458
33478 ?        Ss     0:00 /usr/sbin/haproxy -f /etc/haproxy/haproxy.cfg -p /run/haproxy.pid -Ds -sf 33458
33479 ?        S      0:00 /usr/sbin/haproxy -f /etc/haproxy/haproxy.cfg -p /run/haproxy.pid -Ds
33483 ?        S      0:00 /usr/sbin/haproxy -f /etc/haproxy/haproxy.cfg -p /run/haproxy.pid -Ds -sf 33482
33486 ?        S      0:00 /usr/sbin/haproxy -f /etc/haproxy/haproxy.cfg -p /run/haproxy.pid -Ds -sf 33485
33489 ?        S      0:00 /usr/sbin/haproxy -f /etc/haproxy/haproxy.cfg -p /run/haproxy.pid -Ds -sf 33487
33492 ?        Ss     0:00 /usr/sbin/haproxy -f /etc/haproxy/haproxy.cfg -p /run/haproxy.pid -Ds
33493 ?        Ss     0:00 /usr/sbin/haproxy -f /etc/haproxy/haproxy.cfg -p /run/haproxy.pid -Ds -sf 33482
33499 ?        S      0:00 /usr/sbin/haproxy -f /etc/haproxy/haproxy.cfg -p /run/haproxy.pid -Ds -sf 33498
33500 ?        Ss     0:00 /usr/sbin/haproxy -f /etc/haproxy/haproxy.cfg -p /run/haproxy.pid -Ds -sf 33487
33501 ?        Ss     0:00 /usr/sbin/haproxy -f /etc/haproxy/haproxy.cfg -p /run/haproxy.pid -Ds -sf 33485
33504 ?        S      0:00 /usr/sbin/haproxy -f /etc/haproxy/haproxy.cfg -p /run/haproxy.pid -Ds -sf 33502
33510 ?        S      0:00 /usr/sbin/haproxy -f /etc/haproxy/haproxy.cfg -p /run/haproxy.pid -Ds -sf 33509
33511 ?        Ss     0:00 /usr/sbin/haproxy -f /etc/haproxy/haproxy.cfg -p /run/haproxy.pid -Ds -sf 33498
33515 ?        S      0:00 /usr/sbin/haproxy -f /etc/haproxy/haproxy.cfg -p /run/haproxy.pid -Ds -sf 33512
33519 ?        Ss     0:00 /usr/sbin/haproxy -f /etc/haproxy/haproxy.cfg -p /run/haproxy.pid -Ds -sf 33502
33520 ?        Ss     0:00 /usr/sbin/haproxy -f /etc/haproxy/haproxy.cfg -p /run/haproxy.pid -Ds -sf 33509
33524 ?        S      0:00 /usr/sbin/haproxy -f /etc/haproxy/haproxy.cfg -p /run/haproxy.pid -Ds -sf 33523
33527 ?        Ss     0:00 /usr/sbin/haproxy -f /etc/haproxy/haproxy.cfg -p /run/haproxy.pid -Ds -sf 33512
33529 ?        S      0:00 /usr/sbin/haproxy -f /etc/haproxy/haproxy.cfg -p /run/haproxy.pid -Ds -sf 33528
33530 ?        S      0:00 /usr/sbin/haproxy -f /etc/haproxy/haproxy.cfg -p /run/haproxy.pid -Ds -sf 33528
33538 ?        Ss     0:00 /usr/sbin/haproxy -f /etc/haproxy/haproxy.cfg -p /run/haproxy.pid -Ds -sf 33528
33539 ?        Ss     0:00 /usr/sbin/haproxy -f /etc/haproxy/haproxy.cfg -p /run/haproxy.pid -Ds -sf 33523
33540 ?        Ss     0:00 /usr/sbin/haproxy -f /etc/haproxy/haproxy.cfg -p /run/haproxy.pid -Ds -sf 33528
33541 ?        S      0:00 /usr/sbin/haproxy -f /etc/haproxy/haproxy.cfg -p /run/haproxy.pid -Ds
33551 ?        Ss     0:00 /usr/sbin/haproxy -f /etc/haproxy/haproxy.cfg -p /run/haproxy.pid -Ds
```


The `kill -USR2` is what the systemd unit sends to `haproxy-systemd-wrapper` on reload. So my guess is that haproxy-systemd-wrapper doesn't synchronise the reloads, thereby leading to the same 'parallel reload' situation demonstrated above. This could either be fixed by not using kill in the systemd unit but some blocking command instead or by synchronising the reloads via `haproxy-systemd-wrapper`.

**Comment by tchwpkgorg** [comment]
I'm seeing the same on Ubuntu 14.04 with haproxy 1.6.1-1ppa1~trusty, with the supplied init.d script:

Old processes are never killed; there is no single connection opened to haproxy, so I don't see a reason why the processes would just grow and grow:

1) haproxy was just started:

    root@srv7 ~ # ps aux|grep haprox
    haproxy  14575  0.0  0.0  39728  4876 ?        Ss   03:38   0:00 /usr/sbin/haproxy -f /etc/haproxy/haproxy.cfg -D -p /var/run/haproxy.pid
    haproxy  14576  0.0  0.0  39608  5152 ?        Ss   03:38   0:00 /usr/sbin/haproxy -f /etc/haproxy/haproxy.cfg -D -p /var/run/haproxy.pid
    haproxy  14577  0.0  0.0  39768  5372 ?        Ss   03:38   0:00 /usr/sbin/haproxy -f /etc/haproxy/haproxy.cfg -D -p /var/run/haproxy.pid
    haproxy  14578  0.0  0.0  39872  5460 ?        Ss   03:38   0:00 /usr/sbin/haproxy -f /etc/haproxy/haproxy.cfg -D -p /var/run/haproxy.pid
    haproxy  14579  0.0  0.0  39608  5084 ?        Ss   03:38   0:00 /usr/sbin/haproxy -f /etc/haproxy/haproxy.cfg -D -p /var/run/haproxy.pid
    haproxy  14580  0.0  0.0  39608  5152 ?        Ss   03:38   0:00 /usr/sbin/haproxy -f /etc/haproxy/haproxy.cfg -D -p /var/run/haproxy.pid
    haproxy  14581  0.0  0.0  39592  4876 ?        Ss   03:38   0:00 /usr/sbin/haproxy -f /etc/haproxy/haproxy.cfg -D -p /var/run/haproxy.pid
    haproxy  14582  0.0  0.0  39916  5444 ?        Ss   03:38   0:00 /usr/sbin/haproxy -f /etc/haproxy/haproxy.cfg -D -p /var/run/haproxy.pid


2) let's restart haproxy:

    root@srv7 ~ # /etc/init.d/haproxy restart
     * Restarting haproxy
    [WARNING] 300/034557 (14687) : config : 'stats' statement ignored for frontend 'SFTP_GO_FRONT' as it requires HTTP mode.
    [WARNING] 300/034557 (14687) : config : 'stats' statement ignored for backend 'SFTP_GO_BACK' as it requires HTTP mode.
    [WARNING] 300/034557 (14687) : Proxy 'MAIN_SSL': in multi-process mode, stats will be limited to process assigned to the current request.
    [WARNING] 300/034557 (14687) : Proxy 'MAIN': in multi-process mode, stats will be limited to process assigned to the current request.
    [WARNING] 300/034557 (14687) : Proxy 'GO': in multi-process mode, stats will be limited to process assigned to the current request.
    [WARNING] 300/034557 (14688) : config : 'stats' statement ignored for frontend 'SFTP_GO_FRONT' as it requires HTTP mode.
    [WARNING] 300/034557 (14688) : config : 'stats' statement ignored for backend 'SFTP_GO_BACK' as it requires HTTP mode.
    [WARNING] 300/034557 (14688) : Proxy 'MAIN_SSL': in multi-process mode, stats will be limited to process assigned to the current request.
    [WARNING] 300/034557 (14688) : Proxy 'MAIN': in multi-process mode, stats will be limited to process assigned to the current request.
    [WARNING] 300/034557 (14688) : Proxy 'GO': in multi-process mode, stats will be limited to process assigned to the current request.
    [ OK ]


3) the number of processes has doubled - it will continue to do so with every restart.

    root@srv7 ~ # ps aux|grep haprox
    haproxy  14575  0.0  0.0  39728  4876 ?        Ss   03:38   0:00 /usr/sbin/haproxy -f /etc/haproxy/haproxy.cfg -D -p /var/run/haproxy.pid
    haproxy  14576  0.0  0.0  39608  5152 ?        Ss   03:38   0:00 /usr/sbin/haproxy -f /etc/haproxy/haproxy.cfg -D -p /var/run/haproxy.pid
    haproxy  14577  0.0  0.0  39768  5372 ?        Ss   03:38   0:00 /usr/sbin/haproxy -f /etc/haproxy/haproxy.cfg -D -p /var/run/haproxy.pid
    haproxy  14578  0.0  0.0  39872  5460 ?        Ss   03:38   0:00 /usr/sbin/haproxy -f /etc/haproxy/haproxy.cfg -D -p /var/run/haproxy.pid
    haproxy  14579  0.0  0.0  39608  5084 ?        Ss   03:38   0:00 /usr/sbin/haproxy -f /etc/haproxy/haproxy.cfg -D -p /var/run/haproxy.pid
    haproxy  14580  0.0  0.0  39608  5152 ?        Ss   03:38   0:00 /usr/sbin/haproxy -f /etc/haproxy/haproxy.cfg -D -p /var/run/haproxy.pid
    haproxy  14581  0.0  0.0  39592  4876 ?        Ss   03:38   0:00 /usr/sbin/haproxy -f /etc/haproxy/haproxy.cfg -D -p /var/run/haproxy.pid
    haproxy  14582  0.0  0.0  39916  5444 ?        Ss   03:38   0:00 /usr/sbin/haproxy -f /etc/haproxy/haproxy.cfg -D -p /var/run/haproxy.pid
    haproxy  14689  0.0  0.0  39476  1096 ?        Ss   03:45   0:00 /usr/sbin/haproxy -f /etc/haproxy/haproxy.cfg -D -p /var/run/haproxy.pid
    haproxy  14690  0.0  0.0  39476  1096 ?        Ss   03:45   0:00 /usr/sbin/haproxy -f /etc/haproxy/haproxy.cfg -D -p /var/run/haproxy.pid
    haproxy  14691  0.0  0.0  39476  1096 ?        Ss   03:45   0:00 /usr/sbin/haproxy -f /etc/haproxy/haproxy.cfg -D -p /var/run/haproxy.pid
    haproxy  14692  0.0  0.0  39476  1096 ?        Ss   03:45   0:00 /usr/sbin/haproxy -f /etc/haproxy/haproxy.cfg -D -p /var/run/haproxy.pid
    haproxy  14693  0.0  0.0  39476  1096 ?        Ss   03:45   0:00 /usr/sbin/haproxy -f /etc/haproxy/haproxy.cfg -D -p /var/run/haproxy.pid
    haproxy  14694  0.0  0.0  39476  1096 ?        Ss   03:45   0:00 /usr/sbin/haproxy -f /etc/haproxy/haproxy.cfg -D -p /var/run/haproxy.pid
    haproxy  14695  0.0  0.0  39476  1096 ?        Ss   03:45   0:00 /usr/sbin/haproxy -f /etc/haproxy/haproxy.cfg -D -p /var/run/haproxy.pid
    haproxy  14696  0.0  0.0  39476  1096 ?        Ss   03:45   0:00 /usr/sbin/haproxy -f /etc/haproxy/haproxy.cfg -D -p /var/run/haproxy.pid

**Comment by discordianfish** [comment]
This is what I came up with to solve the issue with my (runit) wrapper script: https://github.com/mesosphere/marathon-lb/commit/598ff09b0c80349418c38c44808a616e77efec74

**Comment by vincentbernat** [comment]
@tchwpkgorg your problem is different as you are restarting HAProxy and not reloading it. It is due to a bug in the init script that doesn't stop the old processes. It should be fixed in 1.6.2-2ppa2~trusty. Could you confirm this is the case?

**Comment by tchwpkgorg** [comment]
@vincentbernat: it is indeed fixed for me with 1.6.2-2ppa2~trusty (though for me, the problem was in restarting, not reloading).

**Comment by lukastribus** [comment]
A couple of related fixes have been committed and are available in latest stable releases 1.5.16 and 1.6.4:

a3aa9e6840fa0bc7a74f3cab7400dcb33ad4537a MINOR: systemd: ensure a reload doesn't mask a stop
6c2f7955e7300fb2cf0596fa5210cdeb3a2aea5f BUG/MINOR: systemd: propagate the correct signal to haproxy
2fadbe5b0a69e03f51d655b7324a176335bd7fdb BUG/MINOR: systemd: report the correct signal in debug message output
1ab5e8642a3537006db8eeed4a3b1bdca2458b33 BUG/MINOR: systemd: ensure we don't miss signals


Those of you still having problems please upgrade to latest stable and report back to the mailing list.

**Comment by phattrent** [comment]
Tried with 1.6.4 on centos 6.7 no dice.  Still have orphaned processes.  In my case it was holding onto old configuration files and I could be sent to the wrong backend pool depending on my luck.  My only workaround was to pkill haproxy and service haproxy start (obviously not ideal since I may lose http packets).

**Comment by wtarreau** [comment]
On Thu, Mar 24, 2016 at 10:55:52AM -0700, phattrent wrote:
> Tried with 1.6.4 on centos 6.7 no dice.  Still have orphaned processes.  In
> my case it was holding onto old configuration files and I could be sent to the
> wrong backend pool depending on my luck.  My only workaround was to pkill
> haproxy and service haproxy start (obviously not ideal since I may lose http
> packets).  

The changes in 1.6.4 concerning orphaned processes are for the systemd wrapper.
Since you're on centos 6 it's unrelated. I guess we'll probably need more info
to reproduce the issue because almost everyone uses the standard reload method
and this problem was never reported yet.

Are you sure the new process has the permissions to send a signal to the old
one ? I don't know maybe there's some selinux or something like this preventing
it from signaling the process it must go. Instead of killing your processes,
you should try to send SIGUSR1 to the old ones to see if they're going away.

Also please check that your pid file has not been removed by whatever we could
imagine. The new processes must have "-sf $oldpids" when you watch them with
"ps auxw". If you don't see the old pids, it means the pid file was empty or
removed and that can explain a lot of things.

Willy

**Comment by sgenchev** [comment]
1.6.4 did not fix it for me on centos7 either. Exact same behavior with fast reloads: multiple haproxy processes running. I use standard systemd service definition and haproxy-systemd-wrapper to restart it.

**Comment by lukastribus** [comment]
Can you follow Willy's suggestions above and tell us what happens?

**Comment by sukrit007** [comment]
Just ran into a similar issue. However,  over period of time when we have greater than 6 haproxy instances,  we start running into situations where old and new haproxy start accepting the traffic. As a result,  we start seeing intermittent  issues with old and new config.

We are using 
```
haproxy -v       
HA-Proxy version 1.6.4 2016/03/13
Copyright 2000-2016 Willy Tarreau <willy@haproxy.org>
```

Haproxy runs inside a docker container with host networking.

We also use keep alive connection so I would expect old haproxy instances to die out when client recycles the connection and creates a new one.  However, even after client resets
- Old haproxy never dies
- Haproxy instances keep piling up even though we are flocking our haproxy reloads.

Here is our global settings for haproxy:

https://github.com/totem/yoda-proxy/blob/develop/etc/confd/templates/haproxy.cfg.tmpl#L78-L88

**Comment by lukastribus** [comment]
Unless people hitting this problem start responding to the questions in this thread and start troubleshooting as per the suggestions above, it won't be possible to fix the issue.

A "me too" along with partial config ain't helping.

**Comment by sgenchev** [comment]
@lukastribus 
More testing:  In my case (nbproc=1) there are always 2 processes running for haproxy. If I send SIGUSR1 to the first process, nothing happens. If I send SIGUSR1 to second process, both processes will die.
In case below: SIGUSR1 to pid 14966 will do nothing
SIGUSR1 to pid 14970 will kill both 14966 and 14970.
2760 ?        Ss     1:15 /usr/sbin/haproxy-systemd-wrapper -f /etc/haproxy/haproxy.cfg -p /run/haproxy.pid
14966 ?        S      0:00 /usr/sbin/haproxy -f /etc/haproxy/haproxy.cfg -p /run/haproxy.pid -Ds -sf 14958
14970 ?        Ss     0:00 /usr/sbin/haproxy -f /etc/haproxy/haproxy.cfg -p /run/haproxy.pid -Ds -sf 14958
30013 ?        S      0:00 /usr/sbin/haproxy -f /etc/haproxy/haproxy.cfg -p /run/haproxy.pid -Ds -sf 30006
30023 ?        Ss     0:06 /usr/sbin/haproxy -f /etc/haproxy/haproxy.cfg -p /run/haproxy.pid -Ds -sf 30006

 PID file is there, and SELinux is disabled.

**Comment by sukrit007** [comment]
@lukastribus   Here is the additional info that might help point towards the problem:

- The new processes have -sf <pid> with them.
- SELinux is not enabled.
- Signal handling is working as expected. In addition, when I kill the process with persistent TCP connection, haproxy removes the old connections. (Which works as I expected)
- We do not use systemd wrapper script for reload.

However,  as the number of haproxy with persistent tcp connections are increasing, we start seeing that old haproxy instances starts accepting the web traffic for new non-persistent http connections and if we remove the old haproxy connections, only new one accepts the connection. Eventually, our haproxy instances go away when client with persistent tcp connection releases the connection,  however not sure why old haproxy instances are still accepting the web traffic. (We only start seeing this after number of haproxy connections have become greater than 5)

**Comment by jorrit** [comment]
Before I upgraded today I used haproxy 1.5.16 on Ubuntu 14.04. I reload haproxy every night. I've been logging the number of processes each day for a couple of days, these are the results:

> april 12th
> haproxy   3914  0.0  0.3  35512  7844 ?        Ss   Apr10   0:52 /usr/sbin/haproxy -f /etc/haproxy/haproxy.cfg -p /var/run/haproxy.pid -D -sf 1262
> haproxy   5864  0.0  0.3  35000  7392 ?        Ss   00:00   0:06 /usr/sbin/haproxy -f /etc/haproxy/haproxy.cfg -p /var/run/haproxy.pid -D -sf 35890
> haproxy   5865  0.0  0.3  35052  7496 ?        Ss   00:00   0:06 /usr/sbin/haproxy -f /etc/haproxy/haproxy.cfg -p /var/run/haproxy.pid -D -sf 35890
> 
> april 13th
> haproxy   3914  0.0  0.3  35512  7844 ?        Ss   Apr10   1:21 /usr/sbin/haproxy -f /etc/haproxy/haproxy.cfg -p /var/run/haproxy.pid -D -sf 1262
> haproxy   5864  0.0  0.3  35284  7620 ?        Ss   Apr12   0:36 /usr/sbin/haproxy -f /etc/haproxy/haproxy.cfg -p /var/run/haproxy.pid -D -sf 35890
> haproxy  39647  0.0  0.3  35104  7448 ?        Ss   00:00   0:12 /usr/sbin/haproxy -f /etc/haproxy/haproxy.cfg -p /var/run/haproxy.pid -D -sf 5865
> haproxy  39648  0.0  0.3  35084  7508 ?        Ss   00:00   0:12 /usr/sbin/haproxy -f /etc/haproxy/haproxy.cfg -p /var/run/haproxy.pid -D -sf 5865
> 
> april 14th
> haproxy   3914  0.0  0.3  35512  7844 ?        Ss   Apr10   1:44 /usr/sbin/haproxy -f /etc/haproxy/haproxy.cfg -p /var/run/haproxy.pid -D -sf 1262
> haproxy   5864  0.0  0.3  35316  7896 ?        Ss   Apr12   1:01 /usr/sbin/haproxy -f /etc/haproxy/haproxy.cfg -p /var/run/haproxy.pid -D -sf 35890
> haproxy  23361  0.0  0.3  35060  7496 ?        Ss   00:00   0:14 /usr/sbin/haproxy -f /etc/haproxy/haproxy.cfg -p /var/run/haproxy.pid -D -sf 39647
> haproxy  23362  0.0  0.3  35088  7500 ?        Ss   00:00   0:14 /usr/sbin/haproxy -f /etc/haproxy/haproxy.cfg -p /var/run/haproxy.pid -D -sf 39647
> haproxy  39648  0.0  0.3  35208  7584 ?        Ss   Apr13   0:35 /usr/sbin/haproxy -f /etc/haproxy/haproxy.cfg -p /var/run/haproxy.pid -D -sf 5865
> 
> april 15th
> haproxy   3914  0.0  0.3  35512  7844 ?        Ss   Apr10   2:02 /usr/sbin/haproxy -f /etc/haproxy/haproxy.cfg -p /var/run/haproxy.pid -D -sf 1262
> haproxy   5864  0.0  0.3  35316  7896 ?        Ss   Apr12   1:19 /usr/sbin/haproxy -f /etc/haproxy/haproxy.cfg -p /var/run/haproxy.pid -D -sf 35890
> haproxy  23361  0.0  0.3  35092  7548 ?        Ss   Apr14   0:32 /usr/sbin/haproxy -f /etc/haproxy/haproxy.cfg -p /var/run/haproxy.pid -D -sf 39647
> haproxy  39648  0.0  0.3  35208  7584 ?        Ss   Apr13   0:52 /usr/sbin/haproxy -f /etc/haproxy/haproxy.cfg -p /var/run/haproxy.pid -D -sf 5865
> haproxy  62765  0.0  0.3  34936  7344 ?        Ss   00:00   0:08 /usr/sbin/haproxy -f /etc/haproxy/haproxy.cfg -p /var/run/haproxy.pid -D -sf 62757

**Comment by Nomon** [comment]
We are also running into this same issue, on both containerized and vms running haproxy, both 1.5 and 1.6, with both -st and -sf.

I just had this happen again, I do not think its related to sending the signal as I just tried to manually to send SIGTERM to the older processes and it has no effect.

**Comment by cbonte** [comment]
@sukrit007 From your configuration and explanations, this is not an issue, but how haproxy is designed according to your configuration.

Your configuration doesn't provide any "timeout http-keep-alive" value. In this case, it will inherit from "timeout client". Here you have keep-alive connections that can live for 615 seconds (more than 10 minutes) before being shut down. If only 1 request arrives on this connection, it will delay the haproxy shutdown for 615 seconds again. And each time a new request is sent to one of those connections, it will keep the old process alive.
I'd recommend to tune your timeouts, first.

**Comment by wtarreau** [comment]
On Mon, Apr 18, 2016 at 07:09:03AM -0700, Matti Savolainen wrote:
> We are also running into this same issue, on both containerized and vms
> running haproxy, both 1.5 and 1.6, with both -st and -sf.
> 
> I just had this happen again, I do not think its related to sending the
> signal as I just tried to manually to send SIGTERM to the older processes and
> it has no effect.

Thanks Matti, that's quite useful information because haproxy doesn't intercept
SIGTERM, so the old process should have died when you issued it! Now we need to
find if one of the libs it's linked against intercepts the signal, but that
still sounds extremely unlikely.

Wait a minute, did you happen to run under systemd ? And if so, did you send
the signal to the haproxy process or to the wrapper ? I'm asking because the
wrapper intercepts SIGTERM.

**Comment by sukrit007** [comment]
@cbonte I agree that until connection is closed haproxy will not die.  However,  my issue is that when I am sending new requests (new TCP connections ), why would old process pick up the requests.  As per the documentation, the new requests should only be picked up new haproxy instance. 

Note: In my tests I generated 100 curl requests (different processes) and about 30 of them went to old haproxy instance when there were 7 haproxy instances.

**Comment by wtarreau** [comment]
On Mon, Apr 18, 2016 at 10:43:40AM -0700, Sukrit Khera wrote:
> @cbonte I agree that until connection is closed haproxy will not die.
> However,  my issue is that when I am sending new requests (new TCP
> connections ), why would old process pick up the requests.  As per the
> documentation, the new requests should only be picked up new haproxy
> instance. 
> 
> Note: In my tests I generated 100 curl requests (different processes) and
> about 30 of them went to old haproxy instance when there were 7 haproxy
> instances.

For me the only explanation is that the old process never received the soft
stop signal (SIGUSR1) which is the one causing an immediate end to the
listening ports.

Now why this signal was not delivered is another story.

**Comment by Nomon** [comment]
@wtarreau directly to the haproxy processes, I sent it to all of them and none of them exited. Also verified that SIGTERM on a freshly restarted process worked correctly after ending the processes with SIGKILL.

**Comment by wtarreau** [comment]
On Mon, Apr 18, 2016 at 11:33:25AM -0700, Matti Savolainen wrote:
> @wtarreau directly to the haproxy processes, I sent it to all of them and
> none of them exited. Also verified that SIGTERM on a freshly restarted
> process worked correctly after ending the processes with SIGKILL.

Next time it happens, please run "strace -tts200 -o log -p <process_pid>"
before sending the signal then send it. I'm interested in discovering what
the process is doing by this time. There are no uninterruptible operations
in haproxy and since it doesn't catch SIGTERM, there's absolutely no valid
reason it wouldn't die. At this point, nothing comes to my mind about what
could cause this unfortunately.

**Comment by lukastribus** [comment]
Are you guys using nbproc > 1? The signal has to be send to all processes. Otherwise you will only kill one of possibly a lot of PIDs, and PID number in netstat may only increase by one, which make it look like "its still there".

So make sure that whatever you pass to -sf/-st or kill is the same as ´pidof haproxy´. And that the content of the pidfile is the same as ´pidof haproxy´.

@sukrit007: if you see this, you can you provide the output of ´sudo ss -lnpt | grep haproxy´. Also what is your nbproc configuration?

**Comment by sukrit007** [comment]
@lukastribus I have not specified nbproc so it defaults to 1. 

I could not filter the output by haproxy,  but my haproxy is proxying following ports:
- 80, 443, 5673, 8081(local only)

Here is the full output for the command for these ports ( ss -lntp |  grep -E 'State|:443|:80|:8081|:5673'):
```
State      Recv-Q Send-Q        Local Address:Port          Peer Address:Port 
LISTEN     0      128                       *:443                         *:*     
LISTEN     0      128                       *:5673                       *:*     
LISTEN     0      128                       *:80                           *:*     
LISTEN     0      128               127.0.0.1:8081                  *:*
```

Currently I have 5 haproxy instances running (for remaining 4 there is active TCP connection on port 5673)

```
ps -efl | grep haproxy
1 S haproxy  14101     1  0  80   0 - 13878 -      Apr18 ?        00:00:00 /usr/sbin/haproxy -f /etc/haproxy/haproxy.cfg -p /var/run/haproxy.pid -D -sf 14080
1 S haproxy  15944     1  0  80   0 - 14049 -      Apr14 ?        00:02:07 /usr/sbin/haproxy -f /etc/haproxy/haproxy.cfg -p /var/run/haproxy.pid -D -sf 15631
1 S haproxy  15980     1  0  80   0 - 14041 -      Apr14 ?        00:01:52 /usr/sbin/haproxy -f /etc/haproxy/haproxy.cfg -p /var/run/haproxy.pid -D -sf 15961
1 S haproxy  18128     1  0  80   0 - 14003 -      Apr18 ?        00:00:01 /usr/sbin/haproxy -f /etc/haproxy/haproxy.cfg -p /var/run/haproxy.pid -D -sf 18107
1 S haproxy  18860     1  0  80   0 - 13906 -      00:31 ?        00:00:02 /usr/sbin/haproxy -f /etc/haproxy/haproxy.cfg -p /var/run/haproxy.pid -D -sf 18843
```

**Comment by Nomon** [comment]
@wtarreau I found the issue and it was not haproxy related, resolved by recompiling the init process spawning the haproxy with go 1.6.1, on both the VM running haproxies and containers running haproxies we used go 1.5 compiled processes to do the actual starting of the haproxy processes so both were effected by the same bug.

https://github.com/golang/go/issues/13164
https://github.com/golang/go/commit/70c9a8187a2732f969aaaa78e0e6f139d04ef3eb
https://github.com/hashicorp/consul-template/issues/442

**Comment by wtarreau** [comment]
On Tue, Apr 19, 2016 at 12:21:30AM -0700, Matti Savolainen wrote:
> @wtarreau I found the issue and it was not haproxy related, resolved by recompiling the init process spawning the haproxy with go 1.6.1, on both the VM running haproxies and containers running haproxies we used go 1.5 compiled processes to do the actual starting of the haproxy processes so both were effected by the same bug.
> 
> https://github.com/golang/go/issues/13164
> https://github.com/golang/go/commit/70c9a8187a2732f969aaaa78e0e6f139d04ef3eb
> https://github.com/hashicorp/consul-template/issues/442

Wow thank you for the detailed reports. I indeed remember having read part
of it when it was first reported, but the thread went on and I didn't get
the whole info. Now I see how we can workaround this issue by adding a few
lines to haproxy, and even possibly emit a warning so that users know they're
at risk.

**Comment by Nomon** [comment]
@suktir007 1.6 has it. 1.6.1 is fixed.

**Comment by wtarreau** [comment]
@matti, thanks to your information I could implement and validate a workaround
that I already merged into 1.7-dev :

   d50b4ac ("MEDIUM: unblock signals on startup.")

I'll backport this into 1.6 and 1.5 as well so that users affected by this issue
can definitely get rid of it. Many thanks.

**Comment by tonycapone** [comment]
Hi there. I'm seeing this issue on Ubuntu 14.04. 
Here's the error message that seems to be associated with the orphaned process when I issue a service haproxy reload:

` * Reloading haproxy haproxy                                                                                                                                                                                                                                                    [ALERT] 119/203146 (11877) : sendto logger #1 failed: Resource temporarily unavailable (errno=11)`

**Comment by Mongey** [comment]
I manually back ported the [`d50b4a`](https://github.com/haproxy/haproxy/commit/d50b4ac0d4acb00e0d2386198191ac329e8dbb77)  ("MEDIUM: unblock signals on startup.") patch to 1.6.4 and I'm still seeing this behaviour 

```
 Main PID: 6766 (haproxy-systemd)
   Memory: 9.6M
   CGroup: /system.slice/haproxy.service
           ├─ 6766 /usr/sbin/haproxy-systemd-wrapper -f /etc/haproxy/haproxy.cfg -p /run/haproxy.pid
           ├─ 6774 /usr/sbin/haproxy -f /etc/haproxy/haproxy.cfg -p /run/haproxy.pid -Ds
           ├─ 6780 /usr/sbin/haproxy -f /etc/haproxy/haproxy.cfg -p /run/haproxy.pid -Ds
           ├─25499 /usr/sbin/haproxy -f /etc/haproxy/haproxy.cfg -p /run/haproxy.pid -Ds -sf 22109
           └─25501 /usr/sbin/haproxy -f /etc/haproxy/haproxy.cfg -p /run/haproxy.pid -Ds -sf 22109
```

Running a `kill -USR1 6774` produces the following from  `strace -tts200 -o log -p 6774`
```
23:32:39.689726 wait4(6780, NULL, 0, NULL) = ? ERESTARTSYS (To be restarted if SA_RESTART is set)
23:32:50.907913 --- SIGUSR1 {si_signo=SIGUSR1, si_code=SI_USER, si_pid=19391, si_uid=0} ---
23:32:50.908328 rt_sigaction(SIGUSR1, {0x4be3e8, [USR1], SA_RESTORER|SA_RESTART, 0x7f054b71d670}, {0x4be3e8, [USR1], SA_RESTORER|SA_RESTART, 0x7f054b71d670}, 8) = 0
23:32:50.908442 rt_sigreturn()          = 61
23:32:50.908514 wait4(6780,  <detached ...>
```
I waited a couple of seconds then detached.

Running  a `kill -USR2 6774` produced
```
23:34:14.213250 wait4(6780, NULL, 0, NULL) = ? ERESTARTSYS (To be restarted if SA_RESTART is set)
23:34:22.204212 --- SIGUSR2 {si_signo=SIGUSR2, si_code=SI_USER, si_pid=19391, si_uid=0} ---
23:34:22.204915 +++ killed by SIGUSR2 +++
```

Leaving the `6780` process still around
` ps -eo pid,comm,lstart,etime,time,args | grep haproxy`
```
 6766 haproxy-systemd Fri May 13 14:34:44 2016  4-09:07:42 00:00:03 /usr/sbin/haproxy-systemd-wrapper -f /etc/haproxy/haproxy.cfg -p /run/haproxy.pid
 6780 haproxy         Fri May 13 14:34:44 2016  4-09:07:42 00:19:09 /usr/sbin/haproxy -f /etc/haproxy/haproxy.cfg -p /run/haproxy.pid -Ds
25499 haproxy         Tue May 17 23:03:31 2016       38:55 00:00:00 /usr/sbin/haproxy -f /etc/haproxy/haproxy.cfg -p /run/haproxy.pid -Ds -sf 22109
25501 haproxy         Tue May 17 23:03:31 2016       38:55 00:00:08 /usr/sbin/haproxy -f /etc/haproxy/haproxy.cfg -p /run/haproxy.pid -Ds -sf 22109
```

**Comment by wtarreau** [comment]
Hi,

On Tue, May 17, 2016 at 04:44:31PM -0700, Conor Mongey wrote:
> I manually back ported the [`d50b4a`](https://github.com/haproxy/haproxy/commit/d50b4ac0d4acb00e0d2386198191ac329e8dbb77)  ("MEDIUM: unblock signals on startup.") patch to 1.6.4 and I'm still seeing this behaviour 
> 

Well, first there were many other fixes between 1.6.4 and 1.6.5 so backporting
a single patch will only fix one part of a problem.

>            ?????? 6766 /usr/sbin/haproxy-systemd-wrapper -f /etc/haproxy/haproxy.cfg -p /run/haproxy.pid
>            ?????? 6774 /usr/sbin/haproxy -f /etc/haproxy/haproxy.cfg -p /run/haproxy.pid -Ds
>            ?????? 6780 /usr/sbin/haproxy -f /etc/haproxy/haproxy.cfg -p /run/haproxy.pid -Ds
>            ??????25499 /usr/sbin/haproxy -f /etc/haproxy/haproxy.cfg -p /run/haproxy.pid -Ds -sf 22109
>            ??????25501 /usr/sbin/haproxy -f /etc/haproxy/haproxy.cfg -p /run/haproxy.pid -Ds -sf 22109
> ```
> 
> Running a `kill -USR1 6774` produces the following from  `strace -tts200 -o log -p 6774`
> ```
> 23:32:39.689726 wait4(6780, NULL, 0, NULL) = ? ERESTARTSYS (To be restarted if SA_RESTART is set)
> 23:32:50.907913 --- SIGUSR1 {si_signo=SIGUSR1, si_code=SI_USER, si_pid=19391, si_uid=0} ---
> 23:32:50.908328 rt_sigaction(SIGUSR1, {0x4be3e8, [USR1], SA_RESTORER|SA_RESTART, 0x7f054b71d670}, {0x4be3e8, [USR1], SA_RESTORER|SA_RESTART, 0x7f054b71d670}, 8) = 0
> 23:32:50.908442 rt_sigreturn()          = 61
> 23:32:50.908514 wait4(6780,  <detached ...>

There is nothing wrong here, you only signaled the parent process which
does not handle any traffic. Either you send your signal to all processes
(using killall) and you'll see all processes eventually die, or you send
it to the wrapper only which will forward it to all processes and deal
with the reload by itself.

Willy

**Comment by phattrent** [comment]
Looks good on our QA Machine, Thanks Willy.

[root@QA-HAP40-01 haproxy-1.6.5]# 
[root@QA-HAP40-01 haproxy-1.6.5]# ps -ef | grep haproxy
haproxy  22654     1  0 10:22 ?        00:00:00 /usr/sbin/haproxy -D -f /etc/haproxy/haproxy.cfg -p /var/run/haproxy.pid
root     22664 20914  0 10:22 pts/0    00:00:00 grep haproxy
[root@QA-HAP40-01 haproxy-1.6.5]# 
[root@QA-HAP40-01 haproxy-1.6.5]# 
[root@QA-HAP40-01 haproxy-1.6.5]# 
[root@QA-HAP40-01 haproxy-1.6.5]# service haproxy reload
Reloading haproxy: 
[root@QA-HAP40-01 haproxy-1.6.5]# service haproxy reload
Reloading haproxy: 
[root@QA-HAP40-01 haproxy-1.6.5]# service haproxy reload
Reloading haproxy: 
[root@QA-HAP40-01 haproxy-1.6.5]# service haproxy reload
Reloading haproxy: 
[root@QA-HAP40-01 haproxy-1.6.5]# service haproxy reload
Reloading haproxy: 
[root@QA-HAP40-01 haproxy-1.6.5]# service haproxy reload
Reloading haproxy: 
[root@QA-HAP40-01 haproxy-1.6.5]# service haproxy reload
Reloading haproxy: 
[root@QA-HAP40-01 haproxy-1.6.5]# 
[root@QA-HAP40-01 haproxy-1.6.5]# 
[root@QA-HAP40-01 haproxy-1.6.5]# 
[root@QA-HAP40-01 haproxy-1.6.5]# 
[root@QA-HAP40-01 haproxy-1.6.5]# 
[root@QA-HAP40-01 haproxy-1.6.5]# ps -ef | grep haproxy
haproxy  22786     1  0 10:22 ?        00:00:00 /usr/sbin/haproxy -D -f /etc/haproxy/haproxy.cfg -p /var/run/haproxy.pid -sf 22769
root     22790 20914  0 10:22 pts/0    00:00:00 grep haproxy
